volmigrate/backend/etcd2/etcd2.go

```diff
@@ -0,0 +1,117 @@
+		log.Fatalf("Failed to create etcd client: %s", err)
+	}
+
+	return &Engine{
+		etcdClient: client.NewKeysAPI(etcdClient),
+		prefix:     prefix,
+	}
+}
+
+// Engine is a migration engine for an etcd v2 datastore.
+type Engine struct {
+	etcdClient client.KeysAPI
+	prefix     string
+}
+
+// CurrentSchemaVersion returns the version of the last migration which was successfuly run.
+// If no previous migrations have been run, the schema version is 0.
+// If there's any error besides "key doesn't exist", execution is aborted.
+func (e *Engine) CurrentSchemaVersion() int64 {
+	resp, err := e.etcdClient.Get(context.Background(), path.Join(e.prefix, backend.SchemaVersionKey), nil)
+	if err != nil {
+		if err := errors.EtcdToErrored(err); err != nil && err == errors.NotExists {
+			return 0 // no key = schema version 0
+		}
+
+		log.Fatalf("Unexpected error when looking up schema version: %v\n", err)
+	}
+
+	i, err := strconv.Atoi(resp.Node.Value)
+	if err != nil {
+		log.Fatalf("Got back unexpected schema version data: %v\n", resp.Node.Value)
+	}
+
+	return int64(i)
+}
+
+// CreateDirectory creates a directory at the target path.
+func (e *Engine) CreateDirectory(target string) error {
+	fmt.Println("Creating directory: " + target)
+	if _, err := e.etcdClient.Set(context.Background(), path.Join(e.prefix, target), "", &client.SetOptions{Dir: true}); err != nil {
+		return errored.Errorf("Failed to create directory: %s", err)
+	}
+
+	return nil
+}
+
+// CreateKey will create a key at the target location.
+func (e *Engine) CreateKey(target string, contents []byte) error {
+	fmt.Println("Creating key: " + target)
+	if _, err := e.etcdClient.Set(context.Background(), path.Join(e.prefix, target), string(contents), nil); err != nil {
+		errored.Errorf("Failed to create key: %s", err)
+	}
+
+	return nil
+}
+
+// DeleteDirectory will recursively delete a target directory.
+func (e *Engine) DeleteDirectory(target string) error {
+	fmt.Println("Deleting directory: " + target)
+
+	if _, err := e.etcdClient.Delete(context.Background(), path.Join(e.prefix, target), &client.DeleteOptions{Dir: true, Recursive: true}); err != nil {
+		return errored.Errorf("Failed to delete directory: %s", err)
+	}
+
+	return nil
+}
+
+// DeleteKey will delete the target key if it exists.
+func (e *Engine) DeleteKey(target string) error {
+	fmt.Println("Deleting key: " + target)
+
+	if _, err := e.etcdClient.Delete(context.Background(), path.Join(e.prefix, target), &client.DeleteOptions{Dir: false}); err != nil {
+		return errored.Errorf("Failed to delete key: %s", err)
+	}
+
+	return nil
+}
+
+// Name returns the name of the engine ("etcd2", "etcd3", "consul", etc.)
+func (e *Engine) Name() string {
+	return "etcd2"
+}
+
+// UpdateSchemaVersion records the version number of the latest migration which successfully ran.
+func (e *Engine) UpdateSchemaVersion(version int64) error {
+	fmt.Printf("Updating schema version key to: %d\n", version)
+
+	data := strconv.Itoa(int(version))
+
```

Comment:
This function is only called from one spot (`Migration`'s `Run()` function), so it can only ever execute if a migration was deemed to be necessary, and necessary implies `m.version > currentVersion`.  I can add this as a sanity, though.  Good catch.